### PR TITLE
Glossary

### PR DESCRIPTION
**Pull request by jasylwong**

### CODE REVIEW

**Comment by jasylwong** [comment]
CodeClimate kept blocking my PR due to an [Instance Variable Assumption](https://github.com/troessner/reek/blob/master/docs/Instance-Variable-Assumption.md).

After trying the suggested fixes, I wasn't able to appease it, so followed the guidance in the [README](https://github.com/troessner/reek/blob/master/README.md#working-with-rails) and simply disabled it.

File: .codeclimate.yml [hunk 1]
```diff
@@ -13,6 +13,9 @@ plugins:
       - spec/
   reek:
     enabled: true
+    checks:
```
Comment (Stivaros) [inline .codeclimate.yml]: As discussed before, the Reek README has a better suggestion which you might as well implement in full here

File: README.md [hunk 2]
```diff
@@ -8,13 +8,29 @@ Contributors:
...
+[Maya Chukujama]()
 
 [Objective](#Objective) | [Process](#process) | [Installation](#installation) | [Running the program](#running_the_program)
 
-## Objective
+## Objective(s)
+
+### Stath:
+- Collate all learning resources into one app
+- Implement a product using DigitalOcean
```
Comment (Stivaros) [inline README.md]: We're on AWS Lightsail, not DO :)

May I suggest
> Practice system architecture and infrastructure engineering

File: app/controllers/glossaries_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# controller for glossary and entries
```
Comment (Stivaros) [inline app/controllers/glossaries_controller.rb]: In cases like this, where you are struggling for meaningful documentation it's best to use `#:nodoc:`, which is the appropriate way of saying you have nothing to say :)

File: app/helpers/glossaries_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module GlossariesHelper
```
[the remark this answers is not in the document]
Comment (Stivaros) [reply app/helpers/glossaries_helper.rb]: As per the above, use `#:nodoc:` where you don't really know what your documenation will be. I won't keep commenting on this throughout the PR but please review the rest as well :)

File: app/views/glossaries/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    <%= f.text_field :definition, autofocus: true, autocomplete: "definition" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Update" %>
+  </div>
+
+  <%= link_to 'Cancel', glossaries_path %>
+
+<% end %>
```
Comment (Stivaros) [inline app/views/glossaries/edit.html.erb]: Surprised that rubocop didn't flag the lack of a newline at the end of the file here.

File: spec/create_glossary_entry_helper.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+def create_glossary_entry(glossary)
```
Comment (Stivaros) [inline spec/create_glossary_entry_helper.rb]: Love this, great stuff.

File: spec/factories/glossaries.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :glossary, class: Glossary do
```
Comment (Stivaros) [inline spec/factories/glossaries.rb]: Where the class and the factory name are the same, you do not need the `class: Glossary` syntax. FactoryBot is smart enough to detect this itself.

```suggestion
  factory :glossary do
```

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'sign_up_helper'
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]: If you configure `spec/support/**.*.rb` as a pattern, these helpers should be picked up automatically and you don't need to `require` them.

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+require 'create_glossary_entry_helper'
+
+feature 'Glossary creation' do
+  context 'user can create glossary term' do
+    let(:user) { build(:user) }
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]: We're not testing sign up here and it takes longer to sign up via the UI than it does via the backend. It's quicker to `let(:user) { create(:user) }` and then simply use the `sign_in` helper ([docs here](https://github.com/heartcombo/devise#integration-tests)).

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    let(:glossary) { build(:glossary) }
+
+    scenario 'with valid inputs will save a new glossary entry' do
+      sign_up(user)
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]:
```suggestion
      sign_in user
```

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      create_glossary_entry(glossary)
+      current_term = Glossary.find_by(term: glossary.term)
+      expect(current_path).to eq "/glossaries/#{current_term.id}"
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]: I wouldn't worry too much about checking the path. It's causing you to query the DB and actually has little value. Why does the user care about the URL? They only care about where they are and what they can see. There are better ways of inferring this information.

File: spec/features/user_can_delete_glossary_entries_spec.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+      sign_up(user)
+      create_glossary_entry(glossary)
+      current_term = Glossary.find_by(term: glossary.term)
+      expect(current_path).to eq "/glossaries/#{current_term.id}"
+      click_link 'Delete'
+      expect(current_path).to eq glossaries_path
+      expect(page).not_to have_content glossary.term
+    end
+  end
+end
```
Comment (Stivaros) [inline spec/features/user_can_delete_glossary_entries_spec.rb:17]: Above notes stand here too.

File: spec/features/user_can_delete_glossary_entries_spec.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    let(:glossary) { build(:glossary) }
+
+    scenario 'will delete a new glossary entry' do
+      sign_up(user)
+      create_glossary_entry(glossary)
+      current_term = Glossary.find_by(term: glossary.term)
+      expect(current_path).to eq "/glossaries/#{current_term.id}"
+      click_link 'Delete'
+      expect(current_path).to eq glossaries_path
+      expect(page).not_to have_content glossary.term
```
Comment (Stivaros) [inline spec/features/user_can_delete_glossary_entries_spec.rb]: Perhaps an expectation for a flash notice that says 'Term deleted' or something?

File: spec/features/user_can_see_glossary_spec.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'sign_up_helper'
+
+feature 'Glossary view' do
+  context 'user can access glossary' do
+    let(:user) { build(:user) }
+
```
Comment (Stivaros) [inline spec/features/user_can_see_glossary_spec.rb:9]: I would use a factory to load some glossaries into the database here...

File: spec/features/user_can_see_glossary_spec.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    scenario 'displays glossary page for user' do
+      sign_up(user)
+      click_link 'Glossary'
+      expect(current_path).to eq '/glossaries'
+    end
```
Comment (Stivaros) [inline spec/features/user_can_see_glossary_spec.rb:14]: ...then check that they're visible here.

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'sign_up_helper'
+require 'create_glossary_entry_helper'
+
+feature 'Glossary edit' do
+  context 'user can edit glossary term' do
+    let(:user) { build(:user) }
+    let(:glossary) { build(:glossary) }
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: Again, if we are not testing creation through the UI I would create at this stage...

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    let(:glossary2) { build(:glossary) }
+
+    scenario 'with valid inputs will update a new glossary entry' do
+      sign_up(user)
+      create_glossary_entry(glossary)
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: ...load the existing pages here...

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+      current_term = Glossary.find_by(term: glossary.term)
+      expect(current_path).to eq "/glossaries/#{current_term.id}"
+      click_link 'Edit'
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: ...then do these bits but whack this code into a helper to tidy up the test :)

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+  context 'user can edit glossary term' do
+    let(:user) { build(:user) }
+    let(:glossary) { build(:glossary) }
+    let(:glossary2) { build(:glossary) }
+
+    scenario 'with valid inputs will update a new glossary entry' do
+      sign_up(user)
+      create_glossary_entry(glossary)
+      current_term = Glossary.find_by(term: glossary.term)
+      expect(current_path).to eq "/glossaries/#{current_term.id}"
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: Not necessary

File: spec/sign_up_helper.rb [hunk 20]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+def sign_up(user)
+  visit '/'
```
Comment (Stivaros) [inline spec/sign_up_helper.rb]:
```suggestion
  visit root_path
```

Try to use rails pathing rather than hardcoding values

**Review by Stivaros — CHANGES_REQUESTED** [review]
Really good work, needs a little refactor as there are some rough edges, see comments. If you have any questions or need support let me know

**Comment by jasylwong** [comment]
I've addressed the majority of your feedback, except for a few unresolved ones which I'll look at tomorrow or Friday.

**Comment by jasylwong** [comment]
Addressed all the review comments; just 3 remaining codeclimate issues to fix now.

File: app/controllers/glossaries_controller.rb [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+  def show; end
+
+  def create
+    @glossary = Glossary.new(glossary_params)
+
+    if @glossary.save
+      redirect_to @glossary
+    else
+      flash.now[:alert] = 'Both term and definition must be provided'
```
Comment (Stivaros) [inline app/controllers/glossaries_controller.rb]: This works in this case but it doesn't scale - best to pull this information from the `record.errors.details` or `record.errors.messages` array(s).

File: app/helpers/glossaries_helper.rb [hunk 22]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+# Provides logic for the views of the glossary
+module GlossariesHelper
```
Comment (Stivaros) [inline app/helpers/glossaries_helper.rb]: Just to recap a previous point - if this file is empty there's a question to be asked as to whether it belongs in the repo.

File: app/models/glossary.rb [hunk 23]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+# class for glossary and entries
+class Glossary < ApplicationRecord
+  validates :term, :definition, presence: true
```
Comment (Stivaros) [inline app/models/glossary.rb:5]: Nice, common sense validations :)

File: app/models/glossary.rb [hunk 24]
```diff
@@ -0,0 +1,6 @@
...
+end
```
Comment (Stivaros) [inline app/models/glossary.rb:8]: If only for readability and nothing else, I'm a  big fan of the [Rails attributes API](https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html) and would like us to implement it here.

File: app/views/glossaries/edit.html.erb [hunk 25]
```diff
@@ -0,0 +1,22 @@
+<h1>Glossary</h1>
+<h2>Edit entry</h2>
+
+
+<%= form_for @glossary do |f| %>
+  <div>
+    <%= f.label :term %><br />
+    <%= f.text_field :term, autofocus: true, autocomplete: "term" %>
```
Comment (Stivaros) [inline app/views/glossaries/edit.html.erb]: Ohhh TIL. Nice use of the `autocomplete` attribute. I didn't know about this!

File: app/views/glossaries/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,22 @@
+<h1>Glossary</h1>
+<h2>Add an entry</h2>
+
+
+<%= form_for @glossary do |f| %>
```
Comment (Stivaros) [inline app/views/glossaries/new.html.erb]: This form is similar to the `edit` form and it's a common pattern in rails to put them in a partial called `form` and then simply call `render form`. Some [Rails docs](https://guides.rubyonrails.org/layouts_and_rendering.html#passing-local-variables) which may help explain this better.

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 27]
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require 'glossary_helper'
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]: This can be required elsewhere rather than within each spec individually.

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+  context 'user can create glossary term' do
+    let(:user) { create(:user) }
+    let(:glossary) { build(:glossary) }
+
+    scenario 'with valid inputs will save a new glossary entry' do
+      sign_in user
+      visit root_path
+      create_glossary_entry(glossary)
+      expect(page).to have_content glossary.term
+      expect(page).to have_content glossary.definition
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb:16]: Excellent

File: spec/features/user_can_create_glossary_entries_spec.rb [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+    scenario 'with blank inputs will send a flash message and will not save' do
+      sign_in user
+      visit root_path
+      click_link 'Glossary'
+      click_link 'Create new entry'
+      fill_in 'Term', with: ''
+      fill_in 'Definition', with: ''
+      click_button 'Add entry'
+      expect(page).not_to have_content glossary.term
+      expect(page).to have_content 'Both term and definition must be provided'
```
Comment (Stivaros) [inline spec/features/user_can_create_glossary_entries_spec.rb]: Nice, again though how can we produce this pattern in a way that will scale better?

File: spec/features/user_can_delete_glossary_entries_spec.rb [hunk 30]
```diff
@@ -0,0 +1,17 @@
...
+require 'rails_helper'
+
+feature 'Glossary deletion' do
+  context 'user can delete glossary term' do
+    let(:user) { create(:user) }
+    let(:glossary) { create(:glossary) }
+
+    scenario 'will delete a new glossary entry' do
+      sign_in user
+      visit "/glossaries/#{glossary.id}"
```
Comment (Stivaros) [inline spec/features/user_can_delete_glossary_entries_spec.rb]:
```suggestion
      visit glossaries_path(glossary)
```

File: spec/features/user_can_see_glossary_spec.rb [hunk 31]
```diff
@@ -0,0 +1,17 @@
...
+
+feature 'Glossary view' do
+  context 'user can access glossary' do
+    let(:user) { create(:user) }
+    let(:glossary) { create(:glossary) }
+
+    scenario 'displays glossary page for user' do
+      term_to_check = glossary.term
+      sign_in user
+      visit '/glossaries'
```
Comment (Stivaros) [inline spec/features/user_can_see_glossary_spec.rb]:
```suggestion
      visit glossaries_path
```

File: spec/features/user_can_see_glossary_spec.rb [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+      expect(page).to have_content term_to_check
```
Comment (Stivaros) [inline spec/features/user_can_see_glossary_spec.rb]: I think the use of `term_to_check` is unnecessary and actually makes the test less readable. 'What is the `term_to_check`?', you end up asking at the crucial time.

File: spec/features/user_can_signup_spec.rb [hunk 33]
```diff
@@ -4,15 +4,11 @@
...
+    let(:user) { create(:user) }
 
     scenario 'with valid credentials will redirect the user to the home page' do
-      visit '/'
-      click_link 'Sign up'
-      fill_in 'Email', with: user.email
-      fill_in 'Password', with: user.password
-      fill_in 'Password confirmation', with: user.password_confirmation
-      click_button 'Sign up'
+      sign_in user
```
Comment (Stivaros) [inline spec/features/user_can_signup_spec.rb]: Hmm, this is overly simplified. We _want_ to test the sign up process in this spec! 😁

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 34]
```diff
@@ -0,0 +1,33 @@
...
+
+feature 'Glossary edit' do
+  context 'user can edit glossary term' do
+    let(:user) { create(:user) }
+    let(:glossary) { create(:glossary) }
+    let(:glossary2) { build(:glossary) }
+
+    scenario 'with valid inputs will update a new glossary entry' do
+      sign_in user
+      visit "/glossaries/#{glossary.id}"
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: You know what to do :)

File: spec/features/user_can_update_glossary_entries_spec.rb [hunk 35]
```diff
@@ -0,0 +1,33 @@
...
+      edit_glossary_entry(glossary2)
+      expect(page).to have_content glossary2.term
+      expect(page).to have_content glossary2.definition
+      expect(page).not_to have_content glossary.term
+    end
+
+    scenario 'with blank inputs sends a flash message but does not update' do
+      sign_in user
+      visit "/glossaries/#{glossary.id}"
```
Comment (Stivaros) [inline spec/features/user_can_update_glossary_entries_spec.rb]: .

File: spec/glossary_helper.rb [hunk 36]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+def create_glossary_entry(glossary)
+  click_link 'Glossary'
+  click_link 'Create new entry'
+  fill_in_glossary_form(glossary.term, glossary.definition)
```
Comment (Stivaros) [inline spec/glossary_helper.rb]: Love this helper and specifically love the nested DRY use of `fill_in_glossary_form`

File: spec/models/glossary_spec.rb [hunk 37]
```diff
@@ -0,0 +1,92 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe Glossary, type: :mode do
```
Comment (Stivaros) [inline spec/models/glossary_spec.rb]:
```suggestion
RSpec.describe Glossary, type: :model do
```

File: spec/models/glossary_spec.rb [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+  let(:glossaryMock) { build(:glossary) }
```
Comment (Stivaros) [inline spec/models/glossary_spec.rb]: What is this filthy camelcasing im my beautiful Ruby code? 😂

File: spec/models/glossary_spec.rb [hunk 39]
```diff
@@ -0,0 +1,92 @@
...
+    end
+
+    it 'will mark entries without a definition as invalid' do
+      glossary = Glossary.new(term: glossaryMock.term)
+      glossary.valid?
+      expect(glossary.errors.details[:definition]).to include(
+        a_hash_including(error: :blank)
+      )
+    end
+
```
Comment (Stivaros) [inline spec/models/glossary_spec.rb:40]: Would also mark a valid record as valid - don't forget that we still happy path test in the unit tests!

File: spec/models/glossary_spec.rb [hunk 40]
```diff
@@ -0,0 +1,92 @@
...
+        term: glossaryMock.term,
+        definition: glossaryMock.definition
+      )
+      expect { glossary.save }.to change(Glossary, :count).by(1)
+    end
+  end
+
+  context 'modification' do
+    before(:each) do
+      @glossary = Glossary.create(
```
Comment (Stivaros) [inline spec/models/glossary_spec.rb]: You've changed syntax and are now using instance variables - why did you move away from `let` syntax on top of `FactoryBot.create`?

File: spec/rails_helper.rb [hunk 41]
```diff
@@ -73,4 +73,7 @@
   config.filter_rails_from_backtrace!
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
+
+  # Include IntegrationHelper module for integration tests via devise
+  config.include Devise::Test::IntegrationHelpers, type: :feature
```
Comment (Stivaros) [inline spec/rails_helper.rb:79]: Nice one. Funnily enough, I was failing to implement this earlier today myself and knew I was pulling a major brainfreeze.

**Review by Stivaros — CHANGES_REQUESTED** [review]
Overall this is really good but I've got quite a few annoying niggly "best practice" comments.  Let me know if you have any questions or require clarification on anything.

File: spec/models/glossary_spec.rb (hunk 38 above)
Comment (jasylwong) [reply spec/models/glossary_spec.rb]: Forgive me, oh merciful Ruby mentor! I have JavaScript sinned 😂

File: spec/features/user_can_see_glossary_spec.rb (hunk 32 above)
Comment (jasylwong) [reply spec/features/user_can_see_glossary_spec.rb]: Using the Factory 'create' method alone does not seem to put the glossary entry into the db. I needed to call glossary in the scenario before it actually registers.

File: spec/features/user_can_delete_glossary_entries_spec.rb (hunk 30 above)
Comment (jasylwong) [reply spec/features/user_can_delete_glossary_entries_spec.rb]: This change makes the test fail. Currently trying to figure out why.

File: app/controllers/glossaries_controller.rb (hunk 21 above)
Comment (jasylwong) [reply app/controllers/glossaries_controller.rb]: I have made use of `record.errors.messages`, but would like to check with you if what I've done is the most appropriate way.

File: app/models/glossary.rb (hunk 24 above)
Comment (jasylwong) [reply app/models/glossary.rb:8]: Note to self: Using the attributes API can help allow for user input that doesn't conform to a data type we want, by actively changing it to that format for us if we write a custom method.

**Review by Stivaros — APPROVED** [review]
Thanks for sticking with it, 100% coverage on the PR is the benchmark we want to achieve. Well done!
